packages/core/src/xml-transformations/xslt-lowering-writer.ts

```diff
@@ -0,0 +1,307 @@
+import * as fileSystem from "fs";
+import * as path from "path";
+import {OutputStream} from "../io/stream/output-stream";
+
+import {
+  XmlTransformation,
+  XmlTemplate,
+  xmlMatchIsLiteral,
+  xmlMatchIsClass,
+  XmlRootTemplate,
+  XmlMatch,
+} from "./xslt-model";
+
+import { XmlWriter, XmlStreamWriter } from "../xml-schema/xml-writer";
+
+import { XSLT_LOWERING } from "./xslt-vocabulary";
+
+const xslNamespace = "http://www.w3.org/1999/XSL/Transform";
+
+export async function saveXsltLoweringToDirectory(
+  model: XmlTransformation,
+  directory: string,
+  name: string
+): Promise<void> {
+  if (!fileSystem.existsSync(directory)) {
+    fileSystem.mkdirSync(directory);
+  }
+
+  const outputStream = fileSystem.createWriteStream(
+    path.join(directory, name + ".xsd")
+  );
+
+  const result = new Promise<void>((accept, reject) => {
+    outputStream.on("close", accept);
+    outputStream.on("error", reject);
+  });
+
+  const stream = {
+    write: async (chunk) => {
+      outputStream.write(chunk);
+    },
+  } as OutputStream;
+  await writeXsltLowering(model, stream);
+  
+  outputStream.end();
+
+  return result;
+}
+
+export async function writeXsltLowering(
+  model: XmlTransformation,
+  stream: OutputStream
+): Promise<void> {
+  const writer = new XmlStreamWriter(stream);
+  await writeTransformationBegin(model, writer);
+  await writeSettings(writer);
+  await writeRootTemplates(model, writer);
+  await writeCommonTemplates(writer);
+  await writeTemplates(model, writer);
+  await writeIncludes(model, writer);
+  await writeFinalTemplates(writer);
+  await writeTransformationEnd(writer);
+}
+
+async function writeTransformationBegin(
+  model: XmlTransformation,
+  writer: XmlWriter
+): Promise<void> {
+  await writer.writeXmlDeclaration("1.0", "utf-8");
+  writer.registerNamespace("xsl", xslNamespace);
+  await writer.writeElementBegin("xsl", "stylesheet");
+  await writer.writeNamespaceDeclaration("xsl", xslNamespace);
+  await writer.writeAndRegisterNamespaceDeclaration(
+    "sp", "http://www.w3.org/2005/sparql-results#"
+  );
+  await writer.writeLocalAttributeValue("version", "2.0");
+  
+  if (model.targetNamespacePrefix != null) {
+    await writer.writeAndRegisterNamespaceDeclaration(
+      model.targetNamespacePrefix,
+      model.targetNamespace
+    );
+  }
+}
+
+async function writeSettings(
+  writer: XmlWriter
+): Promise<void> {
+  await writer.writeElementFull("xsl", "output")(async writer => {
```

Comment:
I see there is a lot of awaits here .. can't you move the await into writeElementFull ? As a result you would not need to type them here so extensively.

Reply:
Perhaps; I could design a new writer interface with non-async methods which would buffer the calls and write that all at once at the end.